### PR TITLE
extend fairness test for preprocessing mitigation

### PR DESCRIPTION
**Pull request by taharallouche**
## Description
Hello ! This builds upon what has been done in https://github.com/fairlearn/fairlearn/pull/1442 for the CorrelationRemover.

The testing module is now extendible: we can easily add any dataset, new mitigation algorithm or metric for the test.

## Tests

- [ ] no new tests required
- [x] new tests added
- [ ] existing tests adjusted

## Documentation

- [ ] no documentation changes needed
- [ ] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

### CODE REVIEW

**Comment by taharallouche** [comment]
The [broken pipeline](https://github.com/fairlearn/fairlearn/actions/runs/16517593101/job/46712834539?pr=1567) is not related to the tests I've added so I'm quite confused (and I can't reproduce it locally for now)

**Comment by taharallouche** [comment]
Hi @TamaraAtanasoska :wave: I haven't been following closely what's happening to the CI, does the failure [here](github.com/fairlearn/fairlearn/actions/runs/16517593101/job/46712834539?pr=1567) ring any bells ? It's a bit weird because my PR only removes and adds a test file

**Comment by TamaraAtanasoska** [comment]
hey @taharallouche, this is a new failure from the last few days. last time I checked was the beginning of the week and it was all green, something must have happened in the meantime. this is one of the `scikit-learn` estimator checks. however I can't reproduce it locally too, even if trying to mirror the CI setup as closely as possible.

**Comment by taharallouche** [comment]
Ok I think I figured it out: it's some floating point error in the estimated target and some inner scikit learn check doesn't like it.

image

**Comment by TamaraAtanasoska** [comment]
cool, thank you!! they can be tedious, but at least we know that there are these kinds of edge cases that the estimator checks are covering 🥳 but what is with that doc build omg! we might need a separate PR for that

**Comment by TamaraAtanasoska** [comment]
@taharallouche we seem to have too little memory capacity on CircleCI, this is why the doc build crashes. Some general optimization attempts (and a bunch of other stuff) in WIP #1568, but ultimately I had to remove two deep learning examples for the build to pass. My local machine runs tests, but is too old to really run the doc build, so I've been testing in CI. 
We can optimise the `PyTorch` examples, but I am wondering, can we get a `large` runner on CircleCI? How does this work organisation-wise? cc @riedgar-ms

File: fairlearn/preprocessing/_prototype_representation_learner.py [hunk 1]
```diff
@@ -268,18 +268,15 @@ def _optimize(
...
-            raise RuntimeError("The loss minimization failed.") from optimization_error
+        result: OptimizeResult = minimize(
+            self._objective,
+            x0=x0,
+            bounds=bounds,
+            args=(X, y, sensitive_features),
+            method="L-BFGS-B",
+            tol=self.tol,
+            options={"maxiter": self.max_iter},
+        )
```
Comment (taharallouche) [inline fairlearn/preprocessing/_prototype_representation_learner.py:279]: I think I'll keep this as it was handy for debugging

File: test/unit/preprocessing/test_preprocessing_fairness.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+@dataclass(frozen=True)
+class FairnessComparisonData:
+    sensitive_feature: pd.Series
+    y_true: pd.Series
+    y_pred_with_mitigation: NDArray
+    y_pred_without_mitigation: NDArray
+
+    def bootstrap(self, seed: int) -> FairnessComparisonData:
+        sample_indices = np.random.RandomState(seed=seed).choice(
```
Comment (TamaraAtanasoska) [inline test/unit/preprocessing/test_preprocessing_fairness.py:43]: I remember recently reading that `numpy.random.default_rng` is the new way to do what `RandomState` did as they will support it further unlike the later. Won't change much for now, but maybe will make the future easier compatibility-wise?

**Review by TamaraAtanasoska — APPROVED** [review]
LGTM! So happy you decided to continue working on this 🥳 I just have one small comment. I opted for not adding this to the changelog, but I always doubt that decision :) did we add the previous work you did for the other mitigator?
